src/serialize.rs

```diff
@@ -0,0 +1,992 @@
+                match $int::expect_scalar_dtype(dtype)? {
+                    // Write a signed integer of the correct size
+                    TypeStr { size: $size, endianness, type_kind: $Int, .. } |
+                    TypeStr { size: $size, endianness, type_kind: $DateTime, .. } => {
+                        let swap_byteorder = endianness.requires_swap(Endianness::of_machine());
+                        Ok($int::AnyEndianWriter { swap_byteorder })
+                    },
+                    type_str => Err(DTypeError::bad_scalar("write", type_str, stringify!($int))),
+                }
+            }
+        }
+
+        impl AutoSerialize for $int {
+            fn default_dtype() -> DType {
+                DType::new_scalar(TypeStr::with_auto_endianness($Int, $size, None))
+            }
+        }
+    };
+}
+
+// Needed by the macro: Methods missing from byteorder
+trait ReadSingleByteExt {
+    #[inline(always)] fn read_u8_(bytes: &[u8]) -> u8 { bytes[0] }
+    #[inline(always)] fn read_i8_(bytes: &[u8]) -> i8 { i8::from_ne_bytes([bytes[0]]) }
+}
+
+impl<E: ByteOrder> ReadSingleByteExt for E {}
+
+/// Needed by the macro: Methods modified to take a generic type param
+trait WriteSingleByteExt: WriteBytesExt {
+    #[inline(always)] fn write_u8_<B>(&mut self, value: u8) -> io::Result<()> { self.write_u8(value) }
+    #[inline(always)] fn write_i8_<B>(&mut self, value: i8) -> io::Result<()> { self.write_i8(value) }
+}
+
+impl<W: WriteBytesExt + ?Sized> WriteSingleByteExt for W {}
+
+// `all()` means "true", `any()` means "false". (these get put inside `cfg`)
+impl_integer_serializable! {
+    @iterate
+    meta: [ (main_ty: Int) (date_ty: TimeDelta) ]
+    remaining: [
+        // numpy doesn't support i128
+        [ 8  i64 (size1: any()) read_i64  write_i64 ]
+        [ 4  i32 (size1: any()) read_i32  write_i32 ]
+        [ 2  i16 (size1: any()) read_i16  write_i16 ]
+        [ 1   i8 (size1: all()) read_i8_  write_i8_ ]
+    ]
+}
+
+impl_integer_serializable! {
+    @iterate
+    meta: [ (main_ty: Uint) (date_ty: DateTime) ]
+    remaining: [
+        // numpy doesn't support i128
+        [ 8  u64 (size1: any()) read_u64  write_u64 ]
+        [ 4  u32 (size1: any()) read_u32  write_u32 ]
+        [ 2  u16 (size1: any()) read_u16  write_u16 ]
+        [ 1   u8 (size1: all()) read_u8_  write_u8_ ]
+    ]
+}
+
+// Takes info about each data size, from largest to smallest.
+macro_rules! impl_float_serializable {
+    ( $( [ $size:literal $float:ident $read_float:ident $write_float:ident ] )+ ) => { $(
+        mod $float {
+            use super::*;
+
+            pub struct AnyEndianReader { pub(super) swap_byteorder: bool }
+            pub struct AnyEndianWriter { pub(super) swap_byteorder: bool }
+
+            #[inline]
+            fn maybe_swap(swap: bool, x: $float) -> $float {
+                match swap {
+                    true => $float::from_bits(x.to_bits().to_be().to_le()),
```

Comment:
Equivalent to `swap_bytes()`?
```suggestion
                    true => $float::from_bits(x.to_bits().swap_bytes()),
```